### ISSUE TITLE
Predictive back handling of multiple `PopScope` in the same `ModalRoute`

### ISSUE THREAD

**Issue by moffatman** [flutter/flutter#144074]
I have multiple widgets which may want to handle the back button by popping their own navigator-like things instead of popping the current route. I noticed when migrating from `WillPopScope` to `PopScope`, if there's more than one within the same `ModalRoute`, it causes a bit of a problem. Since if at least one has `canPop: false`, they will all receive `onPopInvoked(false)`. Therefore firstly, I need to be careful to recheck state in my `onPopInvoked`, I can't assume a state which caused `canPop: false`. And more importantly, if two Widgets both want to handle the back press, they will both do their actions, instead of just the first one.

For now I have corrected it locally by changing the default implementation of `ModalRoute.onPopInvoked to the below`. But I'm not sure about all the use cases (I don't really know the purpose of listening for successful pop).

```dart
  @override
  void onPopInvoked(bool didPop) {
    for (final PopEntry popEntry in _popEntries) {
      if (!didPop && !popEntry.canPopNotifier.value) {
        // Only the first entry which doesn't want to pop gets the no-pop notification
        popEntry.onPopInvoked?.call(didPop);
        break;
      }
      else if (didPop) {
        // Everyone gets the successful-pop notification
        popEntry.onPopInvoked?.call(didPop);
      }
    }
  }
```

What do you think @justinmc ?

**Comment by clragon** [comment]
I also ran into this problem when I tried to implement simple nested navigators, where nested navigators would pop the enclosing navigator when they have no routes and pops left.
I didnt come up with a solution that seemed easy enough to implement, but your workaround might just do it,
while we hopefully wait for the framework to adjust this.

**Comment by darshankawar** [comment]
maybe related : https://github.com/flutter/flutter/issues/138614

**Comment by justinmc** [comment]
@clragon Have you seen the nested Navigator example at here? https://github.com/flutter/flutter/blob/master/examples/api/lib/widgets/navigator_pop_handler/navigator_pop_handler.0.dart

**Comment by justinmc** [comment]
@moffatman Are you doing some kind of custom routing?

Maybe your onPopInvoked implementation is the behavior that users would expect... I can't think of a good common use case for wanting multiple PopScopes to be called when canPop is false.

There are lower level mechanisms that use a similar pattern to your onPopInvoked to handle system backs, in case you haven't seen them. [WidgetsBinding.handlePopRoute](https://github.com/flutter/flutter/blob/24a792dae105e281962a7e3e4c35cbfd9652ff0a/packages/flutter/lib/src/widgets/binding.dart#L774) receives system pops directly and only notify the first WidgetsBindingObserver to handle it. RootBackButtonDispatcher also uses this.

This pattern always worries me though, because it goes by the order of registration. So the order of _popEntries or _observers might not necessarily be in the same order as your routes or your widget tree if one of them is reregistered.

**Comment by moffatman** [comment]
Nested navigation works okay in the common case, because you only have one `PopScope` per `ModalRoute`. In my case, it is essentially a nested navigator and also a different modal-ish popup within the same `ModalRoute`. You're right, with the current registration system, there isn't any handling of priority at all, it's kind of arbitrary just depending on recent build order. It must have been just good luck it did what made sense in my head the first time I checked.

With WillPopScope, each component didn't have to know about each other. I can just include one or both of them within a `ModalRoute` and it will handle the back button amongst each other properly at event-time. 

I guess I can fix it in my app by pulling the `PopScope` out of each component and feeding the pop event into it. But it is messy, because with `PopScope`, it gets more tedious as you also have to pull the state changes and dependencies out of the child widget too, since `willPop` must be set ahead of time (in comparison with `WillPopScope`, where I could just use a GlobalKey for the child and examine the state at event-time).

**Comment by justinmc** [comment]
Let me know if you end up fixing this in your app and have any more thoughts about how to make this easier.

**Comment by clragon** [comment]
@justinmc 

Thanks for the heads up. I have previously not ever seen this widget and was unaware. 
The [cookbook example](https://docs.flutter.dev/cookbook/effects/nested-nav) was updated from `WillPopScope` to `PopScope` but neither show this widget or what I would call "true" or "intuitive" nesting.
I am glad this exists though and I will experiment if it solves my problem accurately. 

Maybe the cookbook could be updated to also show the other kind of nested navigation.

**Comment by justinmc** [comment]
@clragon Good point. @parlough Do you have any thoughts on the cookbook example mentioned in the previous comment? If you're not the right person feel free to pass this along, just tagging you because I see you made some recent changes there. See the nested Navigator example in the [NavigatorPopHandler docs](https://main-api.flutter.dev/flutter/widgets/NavigatorPopHandler-class.html).

**Comment by parlough** [comment]
@justinmc Thanks for the heads up. I'm not too familiar with that recipe personally, or even navigation. @johnpryan might have some more thoughts.

Do you think the cookbook recipe is still helpful as a whole and just needs some updates, or should be replaced with newer content?

**Comment by clragon** [comment]
@justinmc There is actually still one issue I am experiencing with `NavigatorPopHandler`.
While it seems to correctly enable a nested Navigator to pop itself, it doesnt enable that subtree to _know_ that it can be popped this way. What do I mean by that?

![image](https://github.com/flutter/flutter/assets/11785085/fe9b5db0-77a6-4034-befc-39461418ec6e)

We can demonstrate this really easily by adding an AppBar to the example of `NavigatorPopHandler`.
The AppBar at the root route of the inner Navigator doesn't know that it can pop this enclosing route. 

The AppBar decides whether to show the back button based on the `impliesAppBarDismissal` property of the enclosing Route. 
`impliesAppBarDismissal` is composed of 
1. `hasActiveRouteBelow` which is directly checking the Navigator history. In our nested Navigator, this is false.
2. `_entriesImpliesAppBarDismissal` which checks the number of `LocalHistoryEntry`s there are in this Route.

This behaviour still makes "intuitive" nested Navigators difficult.
To "solve" this, a while back I wrote this: 

```dart

class AppBarDismissalProxy extends StatefulWidget {
  /// Tricks the enclosed Route into thinking it can be popped.
  ///
  /// This is necessary to escape nested navigators.
  /// The enclosing navigator has to appropriately handle the pop.
  const AppBarDismissalProxy({
    super.key,
    this.enabled = true,
    required this.child,
  });

  /// Whether the enclosed Route can be popped.
  final bool enabled;

  /// The widget below this widget in the tree.
  final Widget child;

  @override
  State<AppBarDismissalProxy> createState() => _AppBarDismissalProxyState();
}

class _AppBarDismissalProxyState extends State<AppBarDismissalProxy> {
  LocalHistoryEntry? _entry;

  late final ModalRoute<dynamic> _modalRoute = ModalRoute.of<dynamic>(context)!;

  bool _disposed = false;

  @override
  void initState() {
    super.initState();
    _update();
  }

  void _update() {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (widget.enabled) {
        if (_entry != null) return;
        _disposed = false;
        _entry = LocalHistoryEntry(
          onRemove: () {
            assert(
              _disposed,
              'AppBarDismissalProxy lost its local history entry through a pop action.',
            );
            _entry = null;
          },
        );
        _modalRoute.addLocalHistoryEntry(_entry!);
      } else {
        if (_entry == null) return;
        _disposed = true;
        _modalRoute.removeLocalHistoryEntry(_entry!);
      }
    });
  }

  @override
  void didUpdateWidget(covariant AppBarDismissalProxy oldWidget) {
    if (widget.enabled != oldWidget.enabled) {
      _update();
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    _disposed = true;
    if (_entry != null) {
      _modalRoute.removeLocalHistoryEntry(_entry!);
    }
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => widget.child;
}
```

which is essentially creating a `LocalHistoryEntry` just to trick the root Route inside of the nested Navigator into showing us a back button which is then actually handled by the upper Navigator. We install this in the Root route, and then the nested Navigator has to handle popping itself through its parent. 

However, I consider this pretty janky, unwieldy and I am not entirely sure its stable in all configurations. 

I think this might be no longer related enough to this thread, so perhaps I should move it into its own issue.
